File: shared/web-sdk/src/presence/idle-detector.test.ts
import { describe, expect, it, vi } from "vitest";
import {
  DEFAULT_WARNING_FRACTION,
  type IdleDetectorConfig,
  MINIMUM_IDLE_THRESHOLD_MS,
  createIdleDetector,
} from "./idle-detector.ts";

// ---------------------------------------------------------------------------
// A one-hour default is still too long for a snappy unit test. Use a tight
// synthetic threshold (slightly above the 30 s floor) everywhere. The 30 s
// floor itself is tested via its own dedicated block.
// ---------------------------------------------------------------------------

const BASE_CONFIG: IdleDetectorConfig = {
  idleThresholdMs: 60_000, // 60 s
  // warningThresholdMs omitted → defaults to 90% of idleThresholdMs (54 s).
};

/** Expected warning boundary for BASE_CONFIG: 60_000 * 0.9 = 54_000. */
const BASE_WARNING_MS = BASE_CONFIG.idleThresholdMs * DEFAULT_WARNING_FRACTION;

describe("createIdleDetector — config validation", () => {
  it("rejects idleThresholdMs below the 30 s floor", () => {
    expect(() => createIdleDetector({ idleThresholdMs: MINIMUM_IDLE_THRESHOLD_MS - 1 })).toThrow(/idleThresholdMs/);
  });

  it("accepts idleThresholdMs at exactly the 30 s floor", () => {
    expect(() => createIdleDetector({ idleThresholdMs: MINIMUM_IDLE_THRESHOLD_MS })).not.toThrow();
  });

  it("rejects warningThresholdMs >= idleThresholdMs", () => {
    expect(() => createIdleDetector({ idleThresholdMs: 60_000, warningThresholdMs: 60_000 })).toThrow(
      /warningThresholdMs/,
    );
    expect(() => createIdleDetector({ idleThresholdMs: 60_000, warningThresholdMs: 61_000 })).toThrow(
      /warningThresholdMs/,
    );
  });

  it("rejects negative warningThresholdMs", () => {
    expect(() => createIdleDetector({ idleThresholdMs: 60_000, warningThresholdMs: -1 })).toThrow(/warningThresholdMs/);
  });
});

describe("createIdleDetector — initial state", () => {
  it("starts in active state", () => {
    const det = createIdleDetector(BASE_CONFIG);
    expect(det.snapshot().state).toBe("active");
  });

  it("exposes the configured thresholds in the snapshot", () => {
    const det = createIdleDetector(BASE_CONFIG);
    const snap = det.snapshot();
    expect(snap.idleThresholdMs).toBe(60_000);
    expect(snap.warningThresholdMs).toBe(BASE_WARNING_MS);
  });

  it("starts with all suppression flags cleared", () => {
    const det = createIdleDetector(BASE_CONFIG);
    const snap = det.snapshot();
    expect(snap.cycleActive).toBe(false);
    expect(snap.ttsActive).toBe(false);
    expect(snap.demandStayCount).toBe(0);
  });
});

describe("createIdleDetector — tick transitions (no events)", () => {
  it("stays active before the warning threshold elapses", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tick", nowMs: BASE_WARNING_MS - 1 });
    expect(det.snapshot().state).toBe("active");
  });

  it("transitions active → warning exactly at the warning threshold", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tick", nowMs: BASE_WARNING_MS });
    expect(det.snapshot().state).toBe("warning");
  });

  it("stays warning until the idle threshold elapses", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs - 1 });
    expect(det.snapshot().state).toBe("warning");
  });

  it("transitions warning → idle exactly at the idle threshold", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs });
    expect(det.snapshot().state).toBe("idle");
  });

  it("skips warning when a single tick passes idle threshold directly", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 1_000 });
    expect(det.snapshot().state).toBe("idle");
  });

  it("stays idle on further ticks", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs });
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs * 2 });
    expect(det.snapshot().state).toBe("idle");
  });
});

describe("createIdleDetector — event resets", () => {
  for (const kind of ["interaction", "cycle.start", "cycle.end", "tts.start", "tts.end"] as const) {
    it(`${kind} resets timer from warning back to active`, () => {
      const det = createIdleDetector(BASE_CONFIG);
      det.handle({ kind: "tick", nowMs: BASE_WARNING_MS });
      expect(det.snapshot().state).toBe("warning");
      det.handle({ kind, nowMs: BASE_WARNING_MS });
      expect(det.snapshot().state).toBe("active");
    });

    it(`${kind} resets timer from idle back to active`, () => {
      const det = createIdleDetector(BASE_CONFIG);
      det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs });
      expect(det.snapshot().state).toBe("idle");
      det.handle({ kind, nowMs: BASE_CONFIG.idleThresholdMs });
      expect(det.snapshot().state).toBe("active");
    });

    it(`${kind} re-anchors the timer — idle no longer fires at previous boundary`, () => {
      const det = createIdleDetector(BASE_CONFIG);
      // Reset at t=10s; idle should now fire at 10s + 60s = 70s, not 60s.
      det.handle({ kind, nowMs: 10_000 });
      det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs });
      expect(det.snapshot().state).toBe("active");
      det.handle({ kind: "tick", nowMs: 10_000 + BASE_CONFIG.idleThresholdMs });
      // cycle.start / tts.start leave their flag set, which suppresses idle;
      // the other kinds do not — so we only assert the timer re-anchored,
      // not the state at the post-reset boundary.
      expect(det.snapshot().lastResetAtMs).toBe(10_000);
    });
  }

  it("multiple events in sequence keep active and re-anchor each time", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "interaction", nowMs: 5_000 });
    det.handle({ kind: "tts.end", nowMs: 20_000 });
    det.handle({ kind: "cycle.end", nowMs: 40_000 });
    // 40s + warning (54s) = 94s — still active at 90s.
    det.handle({ kind: "tick", nowMs: 90_000 });
    expect(det.snapshot().state).toBe("active");
    // 40s + 54s = 94s — warning.
    det.handle({ kind: "tick", nowMs: 94_000 });
    expect(det.snapshot().state).toBe("warning");
    // 40s + 60s = 100s — idle.
    det.handle({ kind: "tick", nowMs: 100_000 });
    expect(det.snapshot().state).toBe("idle");
  });
});

describe("createIdleDetector — cycle suppression", () => {
  it("sets cycleActive on cycle.start and clears on cycle.end", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "cycle.start", nowMs: 0 });
    expect(det.snapshot().cycleActive).toBe(true);
    det.handle({ kind: "cycle.end", nowMs: 1_000 });
    expect(det.snapshot().cycleActive).toBe(false);
  });

  it("blocks idle transition while cycleActive is set", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "cycle.start", nowMs: 0 });
    // Tick well past idle threshold.
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 30_000 });
    // Idle would fire naturally, but suppression clamps to warning.
    expect(det.snapshot().state).toBe("warning");
  });

  it("resumes idle transition after cycle.end clears the flag", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "cycle.start", nowMs: 0 });
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 10_000 });
    expect(det.snapshot().state).toBe("warning");
    // Cycle ends at 100s, re-anchoring the timer.
    det.handle({ kind: "cycle.end", nowMs: 100_000 });
    expect(det.snapshot().state).toBe("active");
    // Tick 60s later — idle fires now that the flag is clear.
    det.handle({ kind: "tick", nowMs: 160_000 });
    expect(det.snapshot().state).toBe("idle");
  });
});

describe("createIdleDetector — tts suppression", () => {
  it("sets ttsActive on tts.start and clears on tts.end", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tts.start", nowMs: 0 });
    expect(det.snapshot().ttsActive).toBe(true);
    det.handle({ kind: "tts.end", nowMs: 1_000 });
    expect(det.snapshot().ttsActive).toBe(false);
  });

  it("blocks idle transition while ttsActive is set", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tts.start", nowMs: 0 });
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 30_000 });
    expect(det.snapshot().state).toBe("warning");
  });

  it("resumes idle transition after tts.end clears the flag", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "tts.start", nowMs: 0 });
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 10_000 });
    expect(det.snapshot().state).toBe("warning");
    det.handle({ kind: "tts.end", nowMs: 100_000 });
    det.handle({ kind: "tick", nowMs: 160_000 });
    expect(det.snapshot().state).toBe("idle");
  });
});

describe("createIdleDetector — independent cycle + tts flags", () => {
  it("still blocks idle while only ttsActive is set (cycle cleared first)", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "cycle.start", nowMs: 0 });
    det.handle({ kind: "tts.start", nowMs: 0 });
    det.handle({ kind: "cycle.end", nowMs: 10_000 });
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 30_000 });
    // Cycle cleared, TTS still active → suppression still holds.
    expect(det.snapshot().state).toBe("warning");
  });

  it("still blocks idle while only cycleActive is set (tts cleared first)", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "cycle.start", nowMs: 0 });
    det.handle({ kind: "tts.start", nowMs: 0 });
    det.handle({ kind: "tts.end", nowMs: 10_000 });
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 30_000 });
    expect(det.snapshot().state).toBe("warning");
  });

  it("resumes idle only after both flags clear", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "cycle.start", nowMs: 0 });
    det.handle({ kind: "tts.start", nowMs: 0 });
    det.handle({ kind: "tts.end", nowMs: 10_000 });
    det.handle({ kind: "cycle.end", nowMs: 20_000 });
    det.handle({ kind: "tick", nowMs: 20_000 + BASE_CONFIG.idleThresholdMs });
    expect(det.snapshot().state).toBe("idle");
  });
});

describe("createIdleDetector — demandStay suppression", () => {
  it("increments demandStayCount on acquire and decrements on release", () => {
    const det = createIdleDetector(BASE_CONFIG);
    const rel = det.acquireDemandStay();
    expect(det.snapshot().demandStayCount).toBe(1);
    rel();
    expect(det.snapshot().demandStayCount).toBe(0);
  });

  it("blocks idle transition while any demand is held", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.acquireDemandStay();
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 30_000 });
    expect(det.snapshot().state).toBe("warning");
  });

  it("resumes idle after the last demand is released", () => {
    const det = createIdleDetector(BASE_CONFIG);
    const rel = det.acquireDemandStay();
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 10_000 });
    expect(det.snapshot().state).toBe("warning");
    rel();
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 20_000 });
    expect(det.snapshot().state).toBe("idle");
  });

  it("requires every parallel demand to release before suppression lifts", () => {
    const det = createIdleDetector(BASE_CONFIG);
    const a = det.acquireDemandStay();
    const b = det.acquireDemandStay();
    expect(det.snapshot().demandStayCount).toBe(2);
    a();
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 10_000 });
    expect(det.snapshot().state).toBe("warning");
    b();
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs + 20_000 });
    expect(det.snapshot().state).toBe("idle");
  });

  it("release is idempotent — calling it twice does not underflow the counter", () => {
    const det = createIdleDetector(BASE_CONFIG);
    const rel = det.acquireDemandStay();
    rel();
    rel();
    expect(det.snapshot().demandStayCount).toBe(0);
  });
});

describe("createIdleDetector — suppressed warning still observable", () => {
  it("enters warning while suppressed, so instrumentation can still fire", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "cycle.start", nowMs: 0 });
    det.handle({ kind: "tick", nowMs: BASE_WARNING_MS });
    // Naturally warning, suppression doesn't downgrade.
    expect(det.snapshot().state).toBe("warning");
  });
});

describe("createIdleDetector — custom warning threshold", () => {
  it("honours an explicit warningThresholdMs instead of the default fraction", () => {
    const det = createIdleDetector({ idleThresholdMs: 60_000, warningThresholdMs: 10_000 });
    det.handle({ kind: "tick", nowMs: 9_999 });
    expect(det.snapshot().state).toBe("active");
    det.handle({ kind: "tick", nowMs: 10_000 });
    expect(det.snapshot().state).toBe("warning");
  });

  it("allows warningThresholdMs=0 — enters warning on the first tick at or after the anchor", () => {
    // With warningThresholdMs=0, elapsed>=0 classifies as warning. The initial
    // state is still 'active' (machine starts there before any event is seen);
    // the very first tick transitions it.
    const det = createIdleDetector({ idleThresholdMs: 60_000, warningThresholdMs: 0 });
    expect(det.snapshot().state).toBe("active");
    det.handle({ kind: "tick", nowMs: 0 });
    expect(det.snapshot().state).toBe("warning");
  });
});

describe("createIdleDetector — onStateChange subscribers", () => {
  it("fires listeners on each transition with the current snapshot", () => {
    const listener = vi.fn();
    const det = createIdleDetector(BASE_CONFIG);
    det.onStateChange(listener);
    det.handle({ kind: "tick", nowMs: BASE_WARNING_MS });
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs });
    det.handle({ kind: "interaction", nowMs: BASE_CONFIG.idleThresholdMs });
    expect(listener).toHaveBeenCalledTimes(3);
    expect(listener.mock.calls[0]?.[0].state).toBe("warning");
    expect(listener.mock.calls[1]?.[0].state).toBe("idle");
    expect(listener.mock.calls[2]?.[0].state).toBe("active");
  });

  it("does not fire on same-state ticks", () => {
    const listener = vi.fn();
    const det = createIdleDetector(BASE_CONFIG);
    det.onStateChange(listener);
    det.handle({ kind: "tick", nowMs: 1_000 });
    det.handle({ kind: "tick", nowMs: 2_000 });
    expect(listener).not.toHaveBeenCalled();
  });

  it("unsubscribe stops future notifications", () => {
    const listener = vi.fn();
    const det = createIdleDetector(BASE_CONFIG);
    const unsub = det.onStateChange(listener);
    det.handle({ kind: "tick", nowMs: BASE_WARNING_MS });
    expect(listener).toHaveBeenCalledTimes(1);
    unsub();
    det.handle({ kind: "tick", nowMs: BASE_CONFIG.idleThresholdMs });
    expect(listener).toHaveBeenCalledTimes(1);
  });
});

describe("createIdleDetector — monotonic clock discipline", () => {
  it("ignores a tick whose nowMs is before the last anchor (non-monotonic)", () => {
    const det = createIdleDetector(BASE_CONFIG);
    det.handle({ kind: "interaction", nowMs: 10_000 });
    // A late / out-of-order tick before the anchor should not retroactively
    // drive the detector into warning/idle.
    det.handle({ kind: "tick", nowMs: 5_000 });
    expect(det.snapshot().state).toBe("active");
  });
});

File: shared/mobile-sdk/src/commonTest/kotlin/io/sentient/mobilesdk/presence/IdleDetectorTest.kt
package io.sentient.mobilesdk.presence

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFails
import kotlin.test.assertFalse
import kotlin.test.assertTrue

// ---------------------------------------------------------------------------
// IdleDetectorTest — port of web-sdk presence/idle-detector.test.ts.
//
// Pins: FSM invariant (active→warning→idle by elapsed time), suppression-clamp
// logic (cycleActive / ttsActive / demandStay block idle → warning),
// event resets (all event kinds re-anchor the timer), demand-stay
// acquire/release counter, 30_000 floor + 0.9 warning-fraction math.
//
// All time is injected (nowMs: Long); no real clock.
// ---------------------------------------------------------------------------

// A one-hour threshold is too long for snappy tests. Use a tight synthetic
// threshold (above the 30 s floor). warningThresholdMs defaults to 90% of
// idleThresholdMs (i.e. 60_000 * 0.9 = 54_000).
private const val BASE_IDLE_MS = 60_000L
private const val BASE_WARNING_MS = (BASE_IDLE_MS * DEFAULT_WARNING_FRACTION).toLong() // 54_000

private fun baseDetector(): IdleDetector =
    createIdleDetector(IdleDetectorConfig(idleThresholdMs = BASE_IDLE_MS))

class IdleDetectorConfigValidationTest {

    @Test
    fun rejects_idleThresholdMs_below_30s_floor() {
        assertFails {
            createIdleDetector(IdleDetectorConfig(idleThresholdMs = MINIMUM_IDLE_THRESHOLD_MS - 1))
        }
    }

    @Test
    fun accepts_idleThresholdMs_at_exactly_30s_floor() {
        // Should not throw.
        createIdleDetector(IdleDetectorConfig(idleThresholdMs = MINIMUM_IDLE_THRESHOLD_MS))
    }

    @Test
    fun rejects_warningThresholdMs_equal_to_idleThresholdMs() {
        assertFails {
            createIdleDetector(
                IdleDetectorConfig(idleThresholdMs = 60_000, warningThresholdMs = 60_000)
            )
        }
    }

    @Test
    fun rejects_warningThresholdMs_above_idleThresholdMs() {
        assertFails {
            createIdleDetector(
                IdleDetectorConfig(idleThresholdMs = 60_000, warningThresholdMs = 61_000)
            )
        }
    }

    @Test
    fun rejects_negative_warningThresholdMs() {
        assertFails {
            createIdleDetector(
                IdleDetectorConfig(idleThresholdMs = 60_000, warningThresholdMs = -1)
            )
        }
    }
}

class IdleDetectorInitialStateTest {

    @Test
    fun starts_in_active_state() {
        assertEquals(IdleDetectorState.ACTIVE, baseDetector().snapshot().state)
    }

    @Test
    fun exposes_configured_thresholds_in_snapshot() {
        val det = baseDetector()
        val snap = det.snapshot()
        assertEquals(BASE_IDLE_MS, snap.idleThresholdMs)
        assertEquals(BASE_WARNING_MS, snap.warningThresholdMs)
    }

    @Test
    fun starts_with_all_suppression_flags_cleared() {
        val snap = baseDetector().snapshot()
        assertFalse(snap.cycleActive)
        assertFalse(snap.ttsActive)
        assertEquals(0, snap.demandStayCount)
    }
}

class IdleDetectorTickTransitionsTest {

    @Test
    fun stays_active_before_warning_threshold() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Tick(BASE_WARNING_MS - 1))
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
    }

    @Test
    fun transitions_active_to_warning_at_warning_threshold() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Tick(BASE_WARNING_MS))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }

    @Test
    fun stays_warning_until_idle_threshold() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS - 1))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }

    @Test
    fun transitions_warning_to_idle_at_idle_threshold() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }

    @Test
    fun skips_warning_when_single_tick_passes_idle_threshold_directly() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 1_000))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }

    @Test
    fun stays_idle_on_further_ticks() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS))
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS * 2))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }
}

class IdleDetectorEventResetsTest {

    private fun verifyResetsFromWarning(event: IdleDetectorEvent) {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Tick(BASE_WARNING_MS))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
        det.handle(event)
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
    }

    private fun verifyResetsFromIdle(event: IdleDetectorEvent) {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
        det.handle(event)
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
    }

    @Test
    fun interaction_resets_from_warning() {
        verifyResetsFromWarning(IdleDetectorEvent.Interaction(BASE_WARNING_MS))
    }

    @Test
    fun interaction_resets_from_idle() {
        verifyResetsFromIdle(IdleDetectorEvent.Interaction(BASE_IDLE_MS))
    }

    @Test
    fun cycle_start_resets_from_warning() {
        verifyResetsFromWarning(IdleDetectorEvent.CycleStart(BASE_WARNING_MS))
    }

    @Test
    fun cycle_start_resets_from_idle() {
        verifyResetsFromIdle(IdleDetectorEvent.CycleStart(BASE_IDLE_MS))
    }

    @Test
    fun cycle_end_resets_from_warning() {
        verifyResetsFromWarning(IdleDetectorEvent.CycleEnd(BASE_WARNING_MS))
    }

    @Test
    fun cycle_end_resets_from_idle() {
        verifyResetsFromIdle(IdleDetectorEvent.CycleEnd(BASE_IDLE_MS))
    }

    @Test
    fun tts_start_resets_from_warning() {
        verifyResetsFromWarning(IdleDetectorEvent.TtsStart(BASE_WARNING_MS))
    }

    @Test
    fun tts_start_resets_from_idle() {
        verifyResetsFromIdle(IdleDetectorEvent.TtsStart(BASE_IDLE_MS))
    }

    @Test
    fun tts_end_resets_from_warning() {
        verifyResetsFromWarning(IdleDetectorEvent.TtsEnd(BASE_WARNING_MS))
    }

    @Test
    fun tts_end_resets_from_idle() {
        verifyResetsFromIdle(IdleDetectorEvent.TtsEnd(BASE_IDLE_MS))
    }

    @Test
    fun multiple_events_in_sequence_re_anchor_timer_each_time() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Interaction(5_000))
        det.handle(IdleDetectorEvent.TtsEnd(20_000))
        det.handle(IdleDetectorEvent.CycleEnd(40_000))
        // anchor=40_000; warning at 40_000+54_000=94_000 → active at 90_000
        det.handle(IdleDetectorEvent.Tick(90_000))
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
        // warning at 94_000
        det.handle(IdleDetectorEvent.Tick(94_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
        // idle at 40_000+60_000=100_000
        det.handle(IdleDetectorEvent.Tick(100_000))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }

    @Test
    fun event_re_anchors_timer_idle_no_longer_fires_at_previous_boundary() {
        val det = baseDetector()
        // Reset at t=10s; idle should fire at 10_000+60_000=70_000, not 60_000.
        det.handle(IdleDetectorEvent.Interaction(10_000))
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS))
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
        assertEquals(10_000L, det.snapshot().lastResetAtMs)
    }
}

class IdleDetectorCycleSuppressionTest {

    @Test
    fun sets_cycleActive_on_cycle_start_and_clears_on_cycle_end() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.CycleStart(0))
        assertTrue(det.snapshot().cycleActive)
        det.handle(IdleDetectorEvent.CycleEnd(1_000))
        assertFalse(det.snapshot().cycleActive)
    }

    @Test
    fun blocks_idle_transition_while_cycleActive() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.CycleStart(0))
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 30_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }

    @Test
    fun resumes_idle_after_cycle_end_clears_flag() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.CycleStart(0))
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 10_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
        det.handle(IdleDetectorEvent.CycleEnd(100_000))
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
        det.handle(IdleDetectorEvent.Tick(160_000))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }
}

class IdleDetectorTtsSuppressionTest {

    @Test
    fun sets_ttsActive_on_tts_start_and_clears_on_tts_end() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.TtsStart(0))
        assertTrue(det.snapshot().ttsActive)
        det.handle(IdleDetectorEvent.TtsEnd(1_000))
        assertFalse(det.snapshot().ttsActive)
    }

    @Test
    fun blocks_idle_transition_while_ttsActive() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.TtsStart(0))
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 30_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }

    @Test
    fun resumes_idle_after_tts_end_clears_flag() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.TtsStart(0))
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 10_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
        det.handle(IdleDetectorEvent.TtsEnd(100_000))
        det.handle(IdleDetectorEvent.Tick(160_000))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }
}

class IdleDetectorIndependentFlagsTest {

    @Test
    fun still_blocks_idle_while_only_ttsActive_is_set_cycle_cleared_first() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.CycleStart(0))
        det.handle(IdleDetectorEvent.TtsStart(0))
        det.handle(IdleDetectorEvent.CycleEnd(10_000))
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 30_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }

    @Test
    fun still_blocks_idle_while_only_cycleActive_is_set_tts_cleared_first() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.CycleStart(0))
        det.handle(IdleDetectorEvent.TtsStart(0))
        det.handle(IdleDetectorEvent.TtsEnd(10_000))
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 30_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }

    @Test
    fun resumes_idle_only_after_both_flags_clear() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.CycleStart(0))
        det.handle(IdleDetectorEvent.TtsStart(0))
        det.handle(IdleDetectorEvent.TtsEnd(10_000))
        det.handle(IdleDetectorEvent.CycleEnd(20_000))
        det.handle(IdleDetectorEvent.Tick(20_000 + BASE_IDLE_MS))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }
}

class IdleDetectorDemandStayTest {

    @Test
    fun increments_demandStayCount_on_acquire_and_decrements_on_release() {
        val det = baseDetector()
        val rel = det.acquireDemandStay()
        assertEquals(1, det.snapshot().demandStayCount)
        rel()
        assertEquals(0, det.snapshot().demandStayCount)
    }

    @Test
    fun blocks_idle_transition_while_any_demand_is_held() {
        val det = baseDetector()
        det.acquireDemandStay()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 30_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }

    @Test
    fun resumes_idle_after_last_demand_released() {
        val det = baseDetector()
        val rel = det.acquireDemandStay()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 10_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
        rel()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 20_000))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }

    @Test
    fun requires_every_parallel_demand_to_release_before_suppression_lifts() {
        val det = baseDetector()
        val a = det.acquireDemandStay()
        val b = det.acquireDemandStay()
        assertEquals(2, det.snapshot().demandStayCount)
        a()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 10_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
        b()
        det.handle(IdleDetectorEvent.Tick(BASE_IDLE_MS + 20_000))
        assertEquals(IdleDetectorState.IDLE, det.snapshot().state)
    }

    @Test
    fun release_is_idempotent_does_not_underflow_counter() {
        val det = baseDetector()
        val rel = det.acquireDemandStay()
        rel()
        rel()
        assertEquals(0, det.snapshot().demandStayCount)
    }
}

class IdleDetectorSuppressedWarningTest {

    @Test
    fun enters_warning_while_suppressed_so_instrumentation_can_fire() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.CycleStart(0))
        det.handle(IdleDetectorEvent.Tick(BASE_WARNING_MS))
        // Naturally warning — suppression does NOT downgrade to active.
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }
}

class IdleDetectorCustomWarningThresholdTest {

    @Test
    fun honours_explicit_warningThresholdMs_instead_of_default_fraction() {
        val det = createIdleDetector(IdleDetectorConfig(idleThresholdMs = 60_000, warningThresholdMs = 10_000))
        det.handle(IdleDetectorEvent.Tick(9_999))
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
        det.handle(IdleDetectorEvent.Tick(10_000))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }

    @Test
    fun zero_warningThresholdMs_transitions_to_warning_on_first_tick_at_anchor() {
        val det = createIdleDetector(IdleDetectorConfig(idleThresholdMs = 60_000, warningThresholdMs = 0))
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
        det.handle(IdleDetectorEvent.Tick(0))
        assertEquals(IdleDetectorState.WARNING, det.snapshot().state)
    }
}

class IdleDetectorMonotonicClockTest {

    @Test
    fun ignores_tick_with_nowMs_before_last_anchor() {
        val det = baseDetector()
        det.handle(IdleDetectorEvent.Interaction(10_000))
        // A late / out-of-order tick before the anchor should not drive the machine.
        det.handle(IdleDetectorEvent.Tick(5_000))
        assertEquals(IdleDetectorState.ACTIVE, det.snapshot().state)
    }
}
